packages/plugins/openapi/src/rest-generator.ts

```diff
@@ -860,7 +860,7 @@ export class RESTfulOpenAPIGenerator extends OpenAPIGeneratorBase {
         // For compound ids each component is also exposed as a separate fields for read operations,
         // but not required for write operations
         const fields =
-            idFields.length > 1 && mode === 'read' ? model.fields : model.fields.filter((f) => !isIdField(f));
+            idFields.length > 1 && mode !== 'update' ? model.fields : model.fields.filter((f) => !isIdField(f));
```

Comment:
I just checked JSON:API spec again, the convention for CREATE with client-generated ID is to include it in the top-level id field. https://jsonapi.org/format/#crud-creating-client-ids

Do you feel we should actually exclude the id fields for create here, and extract it from the top-level id? So we don't need to duplicate them in the payload. Is that actually the current behavior of the RESTAPIHandler?

Reply:
Also I'm starting to think about the "update" case here 😄. You can actually update id fields with an update operation, right?

Reply:
> Do you feel we should actually exclude the id fields for create here, and extract it from the top-level id? So we don't need to duplicate them in the payload. Is that actually the current behavior of the RESTAPIHandler?

This change is made to reflect the implementation of the RESTAPIHandler (and how Prisma works). We could go the other way instead though, and update the RESTAPIHandler. I'll think a bit and discuss with some colleagues who are using this API.

As for the spec I feel like this is a bit ambiguous in this case. Technically the ID is generated on the server based on the fields. The same goes for update - you update the field and the id changes.

Reply:
After having a quick chat I think our position is that the compound id is generated server side and as such this proposed change does not break with JSON:API. I'm however very open to be convinced otherwise.

I'll dig a bit further into the behaviour of the RESTAPIHandler and make sure that this and #1856 doesn't cause any mismatches.

Reply:
I tried the current rest handler implementation , for "create", yes, you'll need to set the compound id fields in the "attributes" object. The top-level "id" field is ignored. I think this is OK (and probably more natural). I can update the docs to reflect this (as it deviates from JSON:API spec). This applies to single-id models as well.

For "update", it behaves similarly. The model id(s) can be updated by setting id fields directly inside the "attributes" object. Top-level "id" is ignored.

So it feels like to keep the spec and runtime consistent, we should unconditionally include id fields (regardless compound or not) for "create" and "update". For "read", only include id fields when it's compound?

To be honest, I feel things will be cleaner if we just always include the raw id fields for read and mutation in the rest handler implementation 😂

What do you think @thomassnielsen ?

Reply:
I'm thinking that id should always be present in reads and required in updates, but may not be required in writes, depending on wether the id is generated (auto increment or compound) or provided by the client (typically UUID).

In particular this part of the spec allows for this.

> Every resource object MUST also contain an id member, except when the resource object originates at the client and represents a new resource to be created on the server.

So basically always include id in responses, and require id except if the model has either `@id @default(autoincrement())` or `@@id`.

Reply:
I think from the response point of view we're good. A top-level "id" is already included, and for the case of compound ids, each component id is also included in "attributes".

For "create" and "update", what I meant was (I believe) currently the rest handler isn't fully compliant with the JSON:API spec, in that:

1. It doesn't check if the top-level "id" exists
2. It doesn't use the top-level "id", instead id value is always fetched from "attributes"

So, if you do:

```
POST 
{
  "id": "1",
  "type": "Post",
  "attributes": { "id": 2 }
 }
```

You'll end up creating a `Post` with id 2 😅.

I think it's probably OK for now. Changing it will be a breaking change.

Reply:
Your current change specifically checks `mode !== 'update'`, which means for compound id models, their update request input will not have the id component fields. There won't be a way to update the id fields then, right?

Reply:
Yes for reads the top level id should work according to the spec for models that don't generate ids. That's a bug but as you say also a breaking change. Should we create an issue to keep track of this?

> Your current change specifically checks mode !== 'update', which means for compound id models, their update request input will not have the id component fields. There won't be a way to update the id fields then, right?

Yes that wasn't the right solution. I removed the mode !== 'update' check in the latest commit.